io/io/inc/ROOT/TBufferMerger.hxx

```diff
@@ -107,14 +151,18 @@ private:
 
    void Init(std::unique_ptr<TFile>);
 
+   void MergeImpl();
+
    void Merge();
    void Push(TBufferFile *buffer);
+   bool TryMerge(TBufferMergerFile *memfile);
 
+   bool fCompressTemporaryKeys{false};                           //< Enable compression of the TKeys in the TMemFile (save memory at the expense of time, end result is unchanged)
    size_t fAutoSave{0};                                          //< AutoSave only every fAutoSave bytes
-   size_t fBuffered{0};                                          //< Number of bytes currently buffered
+   std::atomic<size_t> fBuffered{0};                             //< Number of bytes currently buffered
```

Comment:
Yes, but it needs to be read by the user (of course out of the loop) to get an indication that the queue is "full".  Checking this value is currently the way for user to avoid the queue from becoming so large that the process run of out memory ... i.e. the user needs to stop producing data :) and wait or doing something else.

Reply:
What I don't understand is why making this atomic changes things for the user, as it's probably still possible to read the value right before modification in any case. Maybe enforcing a maximum limit on buffering internally in `TBufferMerger` rather than exposing this to the user is a better option.

Reply:
The atomic allow the value to be consistent instead of a partial written value.  Whether the value is + or - 1 is likely not essential.

>  Maybe enforcing a maximum limit on buffering internally in TBufferMerger rather than exposing this to the user is a better option.

The problem is what to do once TBufferMerger noticed the exceptional condition.  Rather than "just" blocking, it needs to return/callback to the user so that they can use the resource (the thread that is about to post a TMemFile but can't because the queue is full).  If I understood correctly (eg. from Daniel Riley), this mechanism (checking the queue length) is the mechanism you had negotiated with CMS for CMSSW to use.

Reply:
Yes, indeed, this is the mechanism that was agreed on with CMS. However, let me try to put this another way. In one of your commits, you moved from locking the queue mutex (which I believe is what you probably should keep) to using this atomic. This is more or less equivalent to just removing the lock, as the atomic by itself won't prevent other threads to start pushing more data onto the queue as you read `fBuffered`, so you will read a smaller value than what you should. If you hold the lock, then if some data is already being pushed, you get the buffered size after it's already pushed, not before, and can make a better decision if you push more data or not. If you don't use the lock, several threads can read a small value and push a lot of data at once rather than each deciding on pushing or not based on the real latest value of how much is buffered. I guess this is the reason you added the lock to get the queue size, so either both `GetQueueSize()` and `GetBuffered()` should have the lock, or neither of them should. The way it is now does not make sense to me.

Reply:
I need to review this and change it and add a comment on the reason of the choice .If I remember correctly it was for local efficiency reason and to avoid having to take the lock twice per push to the queue ... but without an in depth look at the usage.

The user code could be either 
(a)
```
if (bufferMerger.GetBuffered() < threshold_in_bytes) // or (bufferMerger.GetQueueSize() < threshold_in_items)
{
     // push to the queue
} else {
     // hold on to the TMemFile and do something else than producing data
     // (alternatively,  continue using the TMemFile and grow it further
}
```
or (b)
```
// push to the queue
if (bufferMerger.GetBuffered() >= threshold_in_bytes) // or (bufferMerger.GetQueueSize() >= threshold_in_items)
{
     // do something else than producing data
}
```
For (b) whether we use the atomic or the lock, since a lot can happen between the push and the if, there would not be a behavior difference.  In all cases the queue can be NumberOfThreads*MemFileSize larger than the threshold as conceivably all N threads could push at the same time on queue that has already N-1 items.

For (a) several threads could reach the `if` statement at the same time and all could then receive the same answer (eg the queue as N-1 items) whether the getter is atomic or use a lock but would then proceed to push to the queue and thus result to the same over-usage ( N items over the threshold ).  [And here in addition the threads that find the queue full needs to hold on to the TMemFile themselves leading to an effective potential over-usage of 2*N items)

What scenario am I missing?

[Note that because the `push to the queue` is indirect, I don't see a way to combine the test and the push.]